### PR TITLE
Improve detection of numbers

### PR DESCRIPTION
**Pull request by pauldambra**
## Changes

see #8488 

The property detection in the UI and in the plugin server used `typeof X == "number"` and so failed for `"95"`

This corrects that (and adds boolean detection since we were writing the tests)

## How did you test this code?

adding unit tests and manually running the migration with several sizes of property definition data sets (see comments)

### CODE REVIEW

**Comment by pauldambra** [comment]
can generate a large property definition table using 

`INSERT INTO posthog_propertydefinition(id, name, is_numerical, volume_30_day, query_usage_30_day, team_id, property_type, property_type_format)
SELECT gen_random_uuid(), md5(RANDOM()::TEXT), false, null, null, 1, 'String', null FROM generate_series(1, 450000);`

That takes about 14-20 seconds to generate 450,000 property definitions that are all strings

Running `time SKIP_SERVICE_VERSION_REQUIREMENTS=1 PRIMARY_DB=clickhouse DEBUG=1 python manage.py migrate` takes about 20 seconds

cc @mariusandra

File: plugin-server/src/worker/ingestion/property-definitions-auto-discovery.ts [hunk 1]
```diff
@@ -75,6 +81,14 @@ export function detectPropertyDefinitionTypes(value: unknown, key: string): Prop
     if (typeof value === 'string') {
         propertyType = PropertyType.String
 
+        if (isNumericString(value)) {
+            propertyType = PropertyType.Numeric
+        }
+
+        if (isBooleanString(value)) {
+            propertyType = PropertyType.Boolean
+        }
```
Comment (mariusandra) [inline plugin-server/src/worker/ingestion/property-definitions-auto-discovery.ts:90]: How do filters work with boolean strings?

**Review by mariusandra — CHANGES_REQUESTED** [review]
Few inline comments, and then I think we should add some quasi-E2E tests that cover the edge cases we discussed in Slack (string greater than comparison with numbers, comparing booleans, etc).

File: plugin-server/src/worker/ingestion/property-definitions-auto-discovery.ts [hunk 2]
```diff
@@ -1,5 +1,9 @@
 import { DateTimePropertyTypeFormat, PropertyType, UnixTimestampPropertyTypeFormat } from '../../types'
 
+export const isNumericString = (candidate: any): boolean => {
+    return !(candidate instanceof Array) && candidate - parseFloat(candidate) + 1 >= 0
```
[the remark this answers is not in the document]
Comment (pauldambra) [reply plugin-server/src/worker/ingestion/property-definitions-auto-discovery.ts:4]: Dang, it is from SO and I had the URL in my buffer

[the remark this answers is not in the document]
Comment (pauldambra) [reply plugin-server/src/worker/ingestion/property-definitions-auto-discovery.ts:7]: comment added

**Comment by pauldambra** [comment]
> some quasi-E2E tests that cover the edge cases we discussed in Slack (string greater than comparison with numbers, comparing booleans, etc).

tests for the property behaviour are mostly in one file e.g. https://github.com/PostHog/posthog/blob/master/ee/clickhouse/models/test/test_property.py#L298  but there are loads in that file

It's hard to assess because there are many integration tests that also pass filters so hard to know how much coverage we have

In this feature we're not changing the querying behaviour of the back-end, or the values we're passing it, but only determining which operator types to show to a user...

**Review by mariusandra — APPROVED** [review]
I think this looks great, but I made one change: removed the migration. It just feels risky...

**Pull request merged**